File: ios/Classes/Data/Result.swift
//
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0
//

import Foundation

struct Result : Codable {
  let isSuccessful: Bool
  let region: BeaconRegion?
  let data: AnyCodable?
  let error: ResultError?
  
  static func success <T: Codable> (with data: T, for region: BeaconRegion? = nil) -> Result {
    return Result(isSuccessful: true, region: region, data: AnyCodable(data), error: nil)
  }
  
  static func failure (of type: ResultError.Kind, message: String? = nil, fatal: Bool? = nil, for region: BeaconRegion? = nil) -> Result {
    return Result(isSuccessful: false, region: region, data: nil, error: ResultError(type: type, message: message, fatal: fatal))
  }
}

struct ResultError: Codable {
  let type: Kind
  let message: String?
  let fatal: Bool?
  
  enum Kind: String, Codable {
    case runtime = "runtime"
    case permissionDenied = "permissionDenied"
    case serviceDisabled = "serviceDisabled"
    case rangingUnavailable = "rangingUnavailable"
    case monitoringUnavailable = "monitoringUnavailable"
  }
}

File: android/src/main/kotlin/io/intheloup/beacons/data/Result.kt
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0

package io.intheloup.beacons.data

data class Result(val isSuccessful: Boolean,
                  val region: RegionModel? = null,
                  val data: Any? = null,
                  val error: Error? = null
) {
    companion object {
        fun success(data: Any, region: RegionModel? = null) = Result(isSuccessful = true, data = data)

        fun failure(type: String, region: RegionModel? = null, message: String? = null, fatal: Boolean = false) = Result(
                isSuccessful = false,
                error = Result.Error(
                        type = type,
                        message = message,
                        fatal = fatal
                )
        )
    }

    data class Error(val type: String,
                     val message: String?,
                     val fatal: Boolean) {

        object Type {
            const val Runtime = "runtime"
            const val LocationNotFound = "locationNotFound"
            const val PermissionDenied = "permissionDenied"
            const val ServiceDisabled = "serviceDisabled"
        }
    }
}